### PR TITLE
Add configuration for gmp under mingw

### PR DESCRIPTION
**Pull request by hhugo**

### CODE REVIEW

File: packages/conf-mingw-w64-gmp-i686/conf-mingw-w64-gmp-i686.1/opam
Comment (zilbuz) [inline packages/conf-mingw-w64-gmp-i686/conf-mingw-w64-gmp-i686.1/opam:10]: Shouldn't you use `gmp` instead of `libgmp` as an argument for `pkgconf`?

File: packages/conf-mingw-w64-gmp-x86_64/conf-mingw-w64-gmp-x86_64.1/opam
Comment (zilbuz) [inline packages/conf-mingw-w64-gmp-x86_64/conf-mingw-w64-gmp-x86_64.1/opam:10]: Shouldn't you use `gmp` instead of `libgmp` as an argument for `pkgconf`?

File: packages/conf-gmp/conf-gmp.4/opam
Comment (zilbuz) [inline packages/conf-gmp/conf-gmp.4/opam:17]: Maybe add a similar test for `os = "win32" & os-distribution = "cygwin"` with `host-arch-x86_64:installed` and `host-arch-x86_32:installed`?
If you use `$(ocamlc -config-var c_compiler)`, you don't even have to add include directories as the MinGW compiler should already use the one where `conf-mingw-...` installed its files.

**Review by zilbuz — COMMENTED** [review]
I've had some issues trying to use your commit to compile zarith with an opam using Cygwin. I'm not sure if this is because of a difference between MSYS2 and Cygwin, but anyway I've added suggestions that solved the issue for me.
I put those changes on a separate commit on my repo if you want to check them out more easily: https://github.com/zilbuz/opam-repository/commit/558a7887fd5e1b631b9a83f595faca1b059f0472

File: packages/conf-mingw-w64-gmp-x86_64/conf-mingw-w64-gmp-x86_64.1/opam
Comment (hhugo) [reply packages/conf-mingw-w64-gmp-x86_64/conf-mingw-w64-gmp-x86_64.1/opam]: fixed

File: packages/conf-mingw-w64-gmp-i686/conf-mingw-w64-gmp-i686.1/opam
Comment (hhugo) [reply packages/conf-mingw-w64-gmp-i686/conf-mingw-w64-gmp-i686.1/opam]: fixed

File: packages/conf-gmp/conf-gmp.4/opam
Comment (hhugo) [reply packages/conf-gmp/conf-gmp.4/opam:17]: fixed

**Comment by hhugo** [comment]
thanks @zilbuz

**Comment by raphael-proust** [comment]
LGTM

Because the windows side is still new, I'd rather someone else also have a look because I don't trust myself to not miss something obvious.

**Comment by hhugo** [comment]
Also, I've put @dra27 as author and maintainer of the new packages as I don't feel competent, he needs to agree to this before merging.

File: packages/conf-mingw-w64-gmp-i686/conf-mingw-w64-gmp-i686.1/opam
Comment (dra27) [inline packages/conf-mingw-w64-gmp-i686/conf-mingw-w64-gmp-i686.1/opam:5]: Very happy to be the maintainer of the opam conf- packages, but...
```suggestion
authors: "Torbjörn Granlund et al"
```
🙂

File: packages/conf-mingw-w64-gmp-x86_64/conf-mingw-w64-gmp-x86_64.1/opam
Comment (dra27) [inline packages/conf-mingw-w64-gmp-x86_64/conf-mingw-w64-gmp-x86_64.1/opam:5]:
```suggestion
authors: "Torbjörn Granlund et al"
```

File: packages/conf-gmp/conf-gmp.4/opam
Comment (dra27) [inline packages/conf-gmp/conf-gmp.4/opam:22]:
```suggestion
    "%{host-arch-x86_64:installed?x86_64:}%%{host-arch-x86_32:installed?i686:}%-w64-mingw32-gcc -c $CFLAGS test.c"
  ] {os = "win32" & os-distribution != "cygwinports"}
```
(as above)

File: packages/conf-gmp-powm-sec/conf-gmp-powm-sec.3/opam
Comment (dra27) [inline packages/conf-gmp-powm-sec/conf-gmp-powm-sec.3/opam:8]: opam-repository-mingw's depext model required the compiler to have been already installed (which worked, as it was using the opam-depext plugin). I'm afraid while it's a noble goal to want to use the same C compiler as ocamlc/ocamlopt, it's not guaranteed to be there.
```suggestion
  ["sh" "-exc" "cc -c $CFLAGS -I/usr/local/include test.c"] {os != "macos" & os != "win32"}
  ["sh" "-exc"
    "%{host-arch-x86_64:installed?x86_64:}%%{host-arch-x86_32:installed?i686:}%-w64-mingw32-gcc -c $CFLAGS test.c"
  ] {os = "win32"}
```

File: packages/conf-gmp-powm-sec/conf-gmp-powm-sec.3/opam
Comment (dra27) [inline packages/conf-gmp-powm-sec/conf-gmp-powm-sec.3/opam:11]: GitHub won't me suggest the change on the correct line (how many years until they'll add this, I wonder....): pedantically, the package ought to refer to the `host-arch-x86_64` etc. as it now uses them:
```diff
-depends: ["conf-gmp"]
+depends: [
+  "conf-gmp"
+  ("host-arch-x86_32" {os = "win32" & os-distribution != "cygwinports"} |
+   "host-arch-x86_64" {os = "win32" & os-distribution != "cygwinports"})
+]
```
It's pedantic because it has them as a transitive dependency.

**Review by dra27 — CHANGES_REQUESTED** [review]
Many thanks for taking this on (and also to @kkirstein for a version of the conf-gmp-powm-sec change in https://github.com/dra27/opam-repository/pull/13).

The main thing is that we can't rely on OCaml in depexts in the same way as opam-repository-mingw did. I've put the various required runes in review commits, but the key thing is that with https://github.com/dra27/opam-repository/commit/ccca603c99e7095473cb14ded4204cefea550add I can now correctly issue:
```
$ opam switch create empty --empty

$ opam list
# Packages matching: installed
# No matches found

$ opam repo add hhugo

$ opam install conf-gmp-powm-sec
The following actions will be performed:
=== install 6 packages
  ∗ conf-gmp                  4 [required by conf-gmp-powm-sec]
  ∗ conf-gmp-powm-sec         3
  ∗ conf-mingw-w64-gcc-x86_64 1 [required by conf-mingw-w64-gmp-x86_64]
  ∗ conf-mingw-w64-gmp-x86_64 1 [required by conf-gmp]
  ∗ conf-pkg-config           3 [required by conf-mingw-w64-gmp-x86_64]
  ∗ host-arch-x86_64          1 [required by conf-gmp-powm-sec]

Proceed with ∗ 6 installations? [y/n]
```
Note that the switch is empty, but it defaults to host-arch-x86_64. If you were then to add a 32-bit Windows compiler, the package would be rebuilt and it would adjust the depexts.

Incidentally, running with `-v`:

```
<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><>  🐫
⬇ retrieved conf-gmp.4  (cached)
⬇ retrieved conf-gmp-powm-sec.3  (cached)
∗ installed host-arch-x86_64.1
+ C:\Users\DRA\AppData\Local\opam\.cygwin\root\bin\x86_64-w64-mingw32-gcc.exe "--version" (CWD=C:\Users\DRA\AppData\Local\opam\empty2\.opam-switch\build\conf-mingw-w64-gcc-x86_64.1)
- x86_64-w64-mingw32-gcc (GCC) 11.4.0
- Copyright (C) 2021 Free Software Foundation, Inc.
- This is free software; see the source for copying conditions.  There is NO
- warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
-
λ compiled  conf-mingw-w64-gcc-x86_64.1
∗ installed conf-mingw-w64-gcc-x86_64.1
+ C:\Users\DRA\AppData\Local\opam\.cygwin\root\bin\pkgconf.exe "--version" (CWD=C:\Users\DRA\AppData\Local\opam\empty2\.opam-switch\build\conf-pkg-config.3)
- 2.2.0
λ compiled  conf-pkg-config.3
∗ installed conf-pkg-config.3
+ C:\Users\DRA\AppData\Local\opam\.cygwin\root\bin\pkgconf.exe "--personality=x86_64-w64-mingw32" "gmp" (CWD=C:\Users\DRA\AppData\Local\opam\empty2\.opam-switch\build\conf-mingw-w64-gmp-x86_64.1)
λ compiled  conf-mingw-w64-gmp-x86_64.1
∗ installed conf-mingw-w64-gmp-x86_64.1
+ C:\Users\DRA\AppData\Local\opam\.cygwin\root\bin\sh.exe "-exc" "x86_64-w64-mingw32-gcc -c $CFLAGS test.c" (CWD=C:\Users\DRA\AppData\Local\opam\empty2\.opam-switch\build\conf-gmp.4)
- + x86_64-w64-mingw32-gcc -c test.c
λ compiled  conf-gmp.4
∗ installed conf-gmp.4
+ C:\Users\DRA\AppData\Local\opam\.cygwin\root\bin\sh.exe "-exc" "x86_64-w64-mingw32-gcc -c $CFLAGS test.c" (CWD=C:\Users\DRA\AppData\Local\opam\empty2\.opam-switch\build\conf-gmp-powm-sec.3)
- + x86_64-w64-mingw32-gcc -c test.c
λ compiled  conf-gmp-powm-sec.3
∗ installed conf-gmp-powm-sec.3
Done.
```

File: packages/conf-gmp-powm-sec/conf-gmp-powm-sec.3/opam
Comment (hhugo) [reply packages/conf-gmp-powm-sec/conf-gmp-powm-sec.3/opam:13]: Applied

**Comment by hhugo** [comment]
@dra27, I've applied all your suggestion.

**Review by dra27 — COMMENTED** [review]
One tiny fix still needed, but otherwise this looks ready to go, thanks!

**Review by dra27 — APPROVED** [review]
Good to go, thanks!

**Comment by hhugo** [comment]
The PR should only affect windows for which the CI is green.
Errors all seem unrelated.
Can we merge ?

**Pull request merged**
